engine/grid.py

```diff
@@ -303,70 +304,106 @@ def find_border_nodes(self):
         cols = self.nodes.shape[1]
         rows = self.nodes.shape[0]
         leftmost_node = None
+        rightmost_node = None
         leftmost_node_pos = None
+        rightmost_node_pos = None
+        
         for row in range(rows):
             for col in range(cols):
                 node = self.nodes[row][col]
+                #border node 1 left active node 
                 if node.is_active and self.is_on_border(row, col, rows-1, cols-1):
                     # check if this is an active node and on the border
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                         leftmost_node = node
                         leftmost_node_pos = (row, col)
+                if node.is_active and self.is_on_border(row, col, rows , cols):
+                    # check if this is an active node and on the border
+                    self.nodes[row][col].is_border = True
+                    if rightmost_node_pos is None or row > rightmost_node_pos[0]:
+                        rightmost_node = node
+                        rightmost_node_pos = (row, col)
+                    
+                    
+                        
         self.border_nodes = border_list
         self.leftmost_node = leftmost_node
         self.leftmost_node_pos = leftmost_node_pos
+        self.rightmost_node = rightmost_node
+        self.rightmost_node_pos = rightmost_node_pos
 
-    # Return bottom most node that is activated in the right column
-
+    
+    ##Return bottom most node that is activated in the right column
     def bottom_rightmost_node(self, pos):
-
         # node_info: (node, row, col)
         candidate_nodes = [node_info for node_info in self.border_nodes if node_info[2] == pos[1]+1]
         if (candidate_nodes == []):
             return None
         else:
             node_info = min(candidate_nodes,key=lambda node_info: node_info[1])
             return (node_info[1],node_info[2])
+    
+    def bottom_leftmost_node(self, pos):
```

Comment:
Let's make sure these have proper documentation; 
def function_name(self, parameter1, parameter2):
   """
   Short description of function
   
   Parameters: 
   parameter1: short description 
   precondition: does parameter 1 have to be a certain value?
   
   parameter2: short description
   precondition: does parameter 2 have to be a certain value?
   
   returns: (or describe procedure if it doesnt return anything)

   """

Reply:
does this look right?
"""
        A helper function for get_all_lawnmower_waypoints_adjustable, finds the leftmost
        node of the row above current pos.

        Parameters:
        parameter1: pos
        precondition: None

        returns:  Returns the furthest left node on the row above the current pos, 
        or return None if there is no nodes in the row above.
        """

Reply:
does this look right?

        """
        A helper function for get_all_lawnmower_waypoints_adjustable, finds the leftmost
        node of the row above current pos.

        Parameters:
        parameter1: pos
        precondition: None

        returns:  Returns the furthest left node on the row above the current pos, 
        or return None if there is no nodes in the row above.
        """